Mathlib/Combinatorics/Graph/Basic.lean

```diff
@@ -0,0 +1,320 @@
+  exists_isLink_ne : ∃ y ≠ x, G.IsLink e x y
+
+@[simp]
+lemma IsNonloopAt.edge_mem (h : G.IsNonloopAt e x) : e ∈ E(G) :=
+  h.inc.edge_mem
+
+@[simp]
+lemma IsNonloopAt.vx_mem (h : G.IsNonloopAt e x) : x ∈ V(G) :=
+  h.inc.vx_mem
+
+lemma IsLoopAt.not_isNonloop_at (h : G.IsLoopAt e x) (y : α) : ¬ G.IsNonloopAt e y := by
+  intro h'
+  obtain ⟨z, hyz, hy⟩ := h'.exists_isLink_ne
+  rw [← h.eq_of_inc hy.inc_left, ← h.eq_of_inc hy.inc_right] at hyz
+  exact hyz rfl
+
+lemma IsNonloopAt.not_isLoopAt (h : G.IsNonloopAt e x) (y : α) : ¬ G.IsLoopAt e y :=
+  fun h' ↦ h'.not_isNonloop_at x h
+
+lemma Inc.isLoopAt_or_isNonloopAt (h : G.Inc e x) : G.IsLoopAt e x ∨ G.IsNonloopAt e x := by
+  obtain ⟨y, hy⟩ := h
+  obtain rfl | hne := eq_or_ne x y
+  · exact .inl hy
+  exact .inr ⟨hy.inc_left, y, hne.symm, hy⟩
+
+/-! ### Adjacency -/
+
+/-- `G.Adj x y` means that `G` has an edge from `x` to `y`. -/
+def Adj (G : Graph α β) (x y : α) : Prop := ∃ e, G.IsLink e x y
+
+lemma Adj.symm (h : G.Adj x y) : G.Adj y x :=
+  ⟨_, h.choose_spec.symm⟩
+
+lemma adj_comm : G.Adj x y ↔ G.Adj y x :=
+  ⟨Adj.symm, Adj.symm⟩
+
+@[simp]
+lemma Adj.mem_left (h : G.Adj x y) : x ∈ V(G) :=
+  h.choose_spec.vx_mem_left
+
+@[simp]
+lemma Adj.mem_right (h : G.Adj x y) : y ∈ V(G) :=
+  h.symm.mem_left
+
+lemma IsLink.adj (h : G.IsLink e x y) : G.Adj x y :=
+  ⟨e, h⟩
+
+/-! ### Extensionality -/
+
+/-- A constructor for `Graph` in which the edge set is inferred from the incidence predicate
+rather than supplied explicitly. -/
+@[simps]
+protected def mk' (vertexSet : Set α) (IsLink : β → α → α → Prop)
+    (isLink_symm : ∀ ⦃e x y⦄, IsLink e x y → IsLink e y x)
+    (eq_or_eq_of_isLink_of_isLink : ∀ ⦃e x y v w⦄, IsLink e x y → IsLink e v w → x = v ∨ x = w)
+    (vx_mem_left_of_isLink : ∀ ⦃e x y⦄, IsLink e x y → x ∈ vertexSet) : Graph α β where
+  vertexSet := vertexSet
+  edgeSet := {e | ∃ x y, IsLink e x y}
+  IsLink := IsLink
+  isLink_symm _ _ _ _ h := isLink_symm h
+  eq_or_eq_of_isLink_of_isLink := eq_or_eq_of_isLink_of_isLink
+  edge_mem_iff_exists_isLink _ := Iff.rfl
+  vx_mem_left_of_isLink := vx_mem_left_of_isLink
+
+@[simp]
+lemma mk'_eq_self (G : Graph α β) : Graph.mk' V(G) G.IsLink (fun _ _ _ ↦ IsLink.symm)
+  (fun _ _ _ _ _ h h' ↦ h.left_eq_or_eq_of_isLink h') (fun _ _ _ ↦ IsLink.vx_mem_left) = G := by
+  have h := G.edgeSet_eq_setOf_exists_isLink
+  cases G with | mk V E IsLink _ _ _ => simpa [Graph.mk'] using h.symm
+
+/-- Two graphs with the same vertex set and binary incidences are equal.
+(We use this as the default extensionality lemma rather than adding `@[ext]`
+to the definition of `Graph`, so it doesn't require equality of the edge sets.) -/
+@[ext]
+protected lemma ext {G₁ G₂ : Graph α β} (hV : V(G₁) = V(G₂))
+    (h : ∀ e x y, G₁.IsLink e x y ↔ G₂.IsLink e x y) : G₁ = G₂ := by
+  rw [← G₁.mk'_eq_self, ← G₂.mk'_eq_self]
+  simp_rw [hV]
+  convert rfl using 2
+  ext
+  rw [h]
+
+/-- Two graphs with the same vertex set and unary incidences are equal. -/
+@[ext]
```

Comment:
We can't have two `@[ext]` lemmas unfortunately. I might suggest the first one  as the `@[ext]` since `IsLink` is basic and avoids an extra existential.